### PR TITLE
Improved null inference

### PR DESCRIPTION
**Pull request by jongardiner**
- Support for making null inferences based off of `if (is_null($a->b->c->d)) ` and several other variants.  
- isset() & empty() language constructs work and (consistent with PHP) do not report an error if a property in the chain is null.
- is_null() will report an error, but will make inferences in the then clause.  The error can be avoided by using `is_null($a?->b?->c?->d)`
- Also, touched up support for is_int,is_string, is_object, is_array.  All of these infer the obvious type when true and infer the opposite (can't possibly be int, string, object, array respectively) on false.

### CODE REVIEW

**Comment by jongardiner** [comment]
I know this is a lot of changes.  I have confirmed that the units pass, Guardrail passes when examining itself, and multiple internal project scan correctly while emitting new errors that weren't formerly detectable.

File: src/TypeComparer.php [hunk 1]
```diff
@@ -253,14 +275,57 @@ static function removeNullOption(ComplexType|Identifier|Name|null $a): ComplexTy
...
+
 	static function removeNamedOption(ComplexType|Identifier|Name|null $a, string $name): ComplexType|Identifier|Name|null {
 		$ret = [];
 		self::forEachType($a, function($el) use ($name, &$ret) {
 			if(!self::isNamedIdentifier($el,$name)) {
 				$ret[]=$el;
 			}
 		});
-		return self::getUniqueTypes(...$ret);
+		$ret2=self::getUniqueTypes(...$ret);
```
Comment (bknutson123) [inline src/TypeComparer.php:327]: You can inline this $ret2 variable

File: src/Evaluators/Expression/CallLike.php [hunk 2]
```diff
@@ -225,6 +234,7 @@ function onStaticCall(Node\Expr\StaticCall $call, SymbolTable $table, ScopeStack
 		}
 		if ($call->class instanceof Node\Name && gettype($call->name) == "string") {
 			$method = $table->getAbstractedMethod(strval($call->class), strval($call->name));
+			echo "looking up ".$call->class."::".$call->name." on ".$call->getLine()."\n";
```
Comment (bknutson123) [inline src/Evaluators/Expression/CallLike.php:237]: Im guessing this was for debugging. Should we remove it?

**Review by bknutson123 — APPROVED** [review]

File: src/Evaluators/Expression/CallLike.php (hunk 2 above)
Comment (jongardiner) [reply src/Evaluators/Expression/CallLike.php]: Removing

**Pull request merged**